docs/website/contents/for-developers/HardForkCombinatorRollbackMonotonicityProposal.md

```diff
@@ -0,0 +1,285 @@
+
+Note the following relations.
+
+- ZEE=Allowed implies ZBE=Allowed.
+- DSR=Insensitive implies QSR=Insensitive.
+
+## Table for the Current Cardano HFC
+
+The following table and associated notes classifies the current HFC implementation in terms of these dimensions.
+
+| DST | CCT | PIL | DSR | ZEE | ZBE | QSR |
+| --- | --- | --- | --- | --- | --- | --- |
+| I   | A   | P 1 | I   | 2   | 3   | I   |
+
+- (1) It's Possible, but only if there's a Chain Growth violation at a particular time during the epoch.
+
+- (2) The current HFC's behavior is inconsistent without making further assumptions.
+  Overall, ZEE is allowed, but it might cause some unexpected behavior.
+  In particular, forecasts always anticipate that there is at least one safe zone of slots in an era (which is not true for eras with zero epochs).
+  That mismatch could manifest as incorrect time-slot translations eg (within ledger rules execution, but not within Node-To-Client query reponses).
+
+- (3) ZBE is allowed when ticking a ledger state; the HFC will invoke multiple translation functions in sequence.
+  But cross-era forecasting cannot skip eras, so no headers after an epoch with zero blocks could be validated, and so blocks could not progatate in the net.
+  Moreover, today's node would not even mint blocks, since the leadership check (at least currently) relies on forecasting.
+
+## Table for an Particular Hypothetical Cardano HFC
+
+The following table and associated notes classifies an hypothetical HFC implementation in terms of these dimensions.
+Nicolas Frisby anticipates this implementation would be easier to understand, less likely to surprise users, more general (ie place fewer requirements on the block types), and not require excessive code changes.
+
+| DST | CCT | PIL | DSR | ZSE | ZBE | QSR |
+| --- | --- | --- | --- | --- | --- | --- |
+| S   | D   | I   | S 2 | D   | D 1 | I 3 |
+
+- (1) ZBE would be achievable for a prefix of eras when constructing the initial HFC ledger state (in the `ProtocolInfo` passed to the Consensus layer), but otherwise the HFC as a block type does not need to support ZBE at all.
+
+- (2) Esgen points out that perhaps DSR could/should still be Insensitive, and that wouldn't obviously lose too much simplicity.
+
+- (3) Having QSR=Insensitive despite DSR=Sensitive requires an additional mechanism.
+  In this design point, that is anticipated to be constraining these queries to only be answered via an immutable ledger state.
+
+## Disclaimer
+
+There may be other points in the design space worth considering.
+(Or even other dimensions, of course.)
+In particular, DST=Insensitive and PIL=Impossible could be achieved by changing the Cardano ledger's governance to match the existing Cardano HFC logic.
+Specifically, if the ledger did not enact governance actions unless there have actually been k+1 blocks since they were ratified even despite possible Chain Growth violations, then the Cardano HFC would never override the Cardano ledger.
```

Comment:
I fail to see the relation between chain growth violations and k + 1 blocks.

Reply:
Right now governance actions are enacted if they both 1) get enough votes and 2) get those votes before a slot-based deadline. The specified choice of _slot_ for the deadline has so far ensured that the final Yes-vs-No status of the governance action would be determined (ie predictable) more than k _blocks_ before the change was actually made (aka "enacted")---unless there was a Chain Growth violation (since Chain Growth ensures at least k blocks of growth per a fixed number of slots).

Reply:
Let me see if I get this: if $b^d_1$ is the block at the voting deadline, where the votes are counted, then in the following chain:

$b^d_1 \rightarrow b^d_2 \rightarrow \ldots \rightarrow b^d_k \rightarrow b^d_{k+1}$

The vote counting is not stable until we reach/add? block $b^d_{k+1}$, right?

And does the insensitivity to ticks stem from the fact that we're counting blocks instead of slots?

Reply:
The deadline is actually a boundary between slots. There's a slot that's the last slot you could sneak in a vote, and the next slot you can't. The boundary between them is the deadline.

If $b^d_1$ is first block immediately before that deadline or anywhere after it, then I think that diagram is correct. Because rolling back up to k blocks would still not allow the votes to change.

---

> The vote counting is not stable until we reach/add? block $b^d_{k+1}$, right?

Hmm. The word "stable" is tricky, since it is often used to refer to something that's one stability window old. But in the presence of CG violations, that could still be subject to rollback.

So it's "actually stable" only once block `k+1` arrives, yes.

> And does the insensitivity to ticks stem from the fact that we're counting blocks instead of slots?

Yep.